app/src/main/java/com/fstyle/structure_android/utils/Validator.java

```diff
@@ -0,0 +1,293 @@
+    private SparseArray<Method> mValidatedMethods;
+    private List<Integer> mMessageInvalids; // list error string res
+    private BaseModel mModelCache;
+
+    @Nullable
+    private Pattern mNGWordPattern;
+
+    public Validator(@ApplicationContext Context context) {
+        if (context instanceof Activity) {
+            throw new ValidationException(
+                    "Context should be get From Application to avoid leak memory");
+        }
+        mContext = context;
+        mErrorTraces = new SparseIntArray();
+        mValidatedMethods = cacheValidatedMethod();
+        mMessageInvalids = new ArrayList<>();
+    }
+
+    private SparseArray<Method> cacheValidatedMethod() {
+        SparseArray<Method> methods = new SparseArray<>();
+        for (Method method : this.getClass().getDeclaredMethods()) {
+            if (method.isAnnotationPresent(ValidMethod.class)) {
+                ValidMethod validMethod = method.getAnnotation(ValidMethod.class);
+                for (int type : validMethod.type()) {
+                    methods.put(type, method);
+                }
+            }
+        }
+        return methods;
+    }
+
+    /**
+     * Check if an object in Validator Model with annotation {@link Optional} is valid
+     */
+    private boolean isValidOptional(Object factor) {
+        if (factor instanceof Integer) {
+            return (Integer) factor == 0;
+        }
+        if (factor instanceof Long) {
+            return (Long) factor == 0;
+        }
+        if (factor instanceof Float) {
+            return (Float) factor == 0.0f;
+        }
+        if (factor instanceof Double) {
+            return (Double) factor == 0.0f;
+        }
+        if (factor instanceof String) {
+            return TextUtils.isEmpty((String) factor);
+        }
+        return (factor instanceof Boolean && !((Boolean) factor)) || factor == null;
+    }
+
+    /**
+     * Validate a field and log the invalid in {@link Validator#mErrorTraces}.
+     * Beside it, log the new invalids which occur in {@link Validator#mMessageInvalids}.
+     */
+    private void validate(Object factor, Rule[] rules, boolean isOptional) {
+        mMessageInvalids.clear();
+        for (Rule rule : rules) {
+            for (int type : rule.types()) {
+                Method method = mValidatedMethods.get(type);
+                if (method == null) {
+                    continue;
+                }
+                try {
+                    boolean isValid =
+                            (isOptional && isValidOptional(factor)) || (boolean) method.invoke(this,
+                                    factor);
+                    if (isValid) {
+                        mErrorTraces.delete(rule.message());
+                    } else {
+                        mErrorTraces.put(type, rule.message());
+                    }
+                } catch (IllegalAccessException e) {
+                    Log.e(TAG, "validate: ", e);
```

Comment:
WARNING: [AvoidDuplicateLiterals] The String literal "validate: " appears 4 times in this file; the first occurrence is on line 119
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/strings.html#AvoidDuplicateLiterals